packages/lab/src/date-picker/DatePickerSingleInput.tsx

```diff
@@ -28,113 +28,167 @@ const withBaseName = makePrefixer("saltDatePickerSingleInput");
-        </Button>
+    const {
+      className,
+      onFocus,
+      onBlur,
+      value: valueProp,
+      validate,
+      defaultValue,
+      onDateValueChange,
+      onKeyDown,
+      ...rest
+    } = props;
+
+    const {
+      state: { selectedDate, disabled, readOnly, cancelled, minDate, maxDate, resetRequired },
+      helpers: { select },
+    } = useDatePickerContext<TDate>({ selectionVariant: "single" });
+    const {
+      state: { open },
+      helpers: { setOpen },
+    } = useDatePickerOverlay();
+
+    const previousValue = useRef<typeof valueProp>();
+
+    const [value, setValue] = useControlled({
+      controlled: valueProp,
+      default: defaultValue,
+      name: "DatePickerSingleInput",
+      state: "value",
+    });
+
+    const handleCalendarButton = useCallback(() => {
+      setOpen(!open);
+    }, [open, setOpen]);
+
+    const handleDateChange = useCallback(
+      (
+        _event: SyntheticEvent,
+        _date: SingleDateSelection<TDate> | null | undefined,
+        details: DateInputSingleDetails<TDate>,
+      ) => {
+        const validatedSelection = validate
+          ? validate(details)
+          : defaultSingleValidation<TDate>(
+              dateAdapter,
+              details,
+              minDate,
+              maxDate,
+            );
+        select(validatedSelection);
+      },
+      [select, validate],
+    );
+
+    const handleDateValueChange = useCallback(
+      (newDateValue: string) => {
+        setValue(newDateValue);
+        onDateValueChange?.(newDateValue);
+      },
+      [onDateValueChange],
+    );
+
+    const handleOnKeyDown = useCallback(
+      (event: KeyboardEvent<HTMLInputElement>) => {
+        if (event.key === "ArrowDown") {
+          setOpen(true);
+          onKeyDown?.(event);
+        }
+      },
+      [onKeyDown],
+    );
+
+    // biome-ignore lint/correctness/useExhaustiveDependencies: should run when open changes and not selected date or value
+    useEffect(() => {
+      if (open) {
+        previousValue.current = value;
       }
-      {...getReferenceProps({
-        ...rest,
-        onKeyDown: (event: KeyboardEvent<HTMLInputElement>) => {
-          if (event.key === "ArrowDown") {
-            setOpen(true);
-          }
-        },
-      })}
-    />
-  );
-});
+    }, [open]);
+
+    // biome-ignore lint/correctness/useExhaustiveDependencies: avoid excessive re-rendering
+    useEffect(() => {
+      if (cancelled) {
+        setValue(previousValue?.current);
+      }
+    }, [cancelled]);
+
+    useEffect(() => {
+      if (resetRequired) {
+        setValue("");
+      }
+    }, [resetRequired]);
+
+    return (
+      <DateInputSingle
+        value={value || ""}
+        className={clsx(withBaseName(), className)}
+        date={selectedDate || null}
+        readOnly={readOnly}
+        ref={ref}
+        onDateChange={handleDateChange}
+        onDateValueChange={handleDateValueChange}
+        endAdornment={
```

Comment:
Should calendar icon button be part of Date input (more like combobox)? e.g., if I pass in a clear button as `endAdornment`, then calendar icon is gone

Reply:
not asking to couple them, could do similar to ComboBox to PillInput, it automatically adds the button for `endAdornment`
https://github.com/jpmorganchase/salt-ds/blob/7d3251821e4b93cdf1eab55a99ec9a5213dfe296/packages/core/src/combo-box/ComboBox.tsx#L415-L419